File: packages/depz-sensor-sdk-java/src/main/java/ai/depz/sensor/protocol/Sr04.java
package ai.depz.sensor.protocol;

/** SR04 wire codecs (contracts/03_SENSOR_SR04.md). */
public final class Sr04 {
    private Sr04() {}

    public enum Sr04Cmd {
        GET_SAMPLE_PERIOD(0x32),
        SET_SAMPLE_PERIOD(0x33),
        GET_ECHO_DECAY(0x34),
        SET_ECHO_DECAY(0x35),
        MEASURE_ONCE(0x36),
        START_MEASUREMENT_LOOP(0x37),
        STOP_MEASUREMENT_LOOP(0x38);

        public final int value;
        Sr04Cmd(int value) { this.value = value; }
    }

    public enum Sr04Rpt {
        DATA(0x91),
        SAMPLE_PERIOD(0x92),
        ECHO_DECAY(0x93);

        public final int value;
        Sr04Rpt(int value) { this.value = value; }
    }

    /** echo_time_us sentinel: no echo received. */
    public static final int ECHO_TIMEOUT = 0xFFFF;

    public static final long SAMPLE_PERIOD_DEFAULT_US = 50_000L;
    public static final int ECHO_DECAY_DEFAULT_US = 5_000;
    public static final int ECHO_DECAY_MIN_US = 4_000;
    public static final int ECHO_DECAY_MAX_US = 65_000;

    /** {@code sourceCmd}: 0x36 single shot (host or SYNC_IN), 0x37 loop sample. */
    public record Sr04Data(int sourceCmd, long timestampUs, int echoTimeUs) {
        public static Sr04Data unpack(byte[] p) {
            int cmd = p[0] & 0xFF;
            long ts = Common.u64le(p, 1);
            int echo = (p[9] & 0xFF) | ((p[10] & 0xFF) << 8);
            return new Sr04Data(cmd, ts, echo);
        }
    }

    public static byte[] packSamplePeriod(long periodUs) {
        return new byte[] {
            (byte) periodUs, (byte) (periodUs >>> 8), (byte) (periodUs >>> 16), (byte) (periodUs >>> 24)
        };
    }

    /** Reads the u32 sample period as an unsigned value in a long. */
    public static long unpackSamplePeriod(byte[] p) {
        return (p[0] & 0xFFL) | ((p[1] & 0xFFL) << 8) | ((p[2] & 0xFFL) << 16) | ((p[3] & 0xFFL) << 24);
    }

    public static byte[] packEchoDecay(int decayUs) {
        return new byte[] {(byte) decayUs, (byte) (decayUs >>> 8)};
    }

    public static int unpackEchoDecay(byte[] p) {
        return (p[0] & 0xFF) | ((p[1] & 0xFF) << 8);
    }

    /**
     * Round-trip echo time → distance in mm; {@code null} for the timeout
     * sentinel. Default speed of sound 343 m/s; with {@code airTempC} uses
     * c = 331.3 + 0.606·T (m/s).
     */
    public static Double distanceMmFromEcho(int echoTimeUs, Double airTempC) {
        if (echoTimeUs == ECHO_TIMEOUT) {
            return null;
        }
        double cMs = (airTempC == null) ? 343.0 : 331.3 + 0.606 * airTempC;
        return echoTimeUs * cMs / 2000.0;
    }
}

File: packages/depz-sensor-sdk-rust/src/protocol/sr04.rs
//! SR04 wire codecs (contracts/03_SENSOR_SR04.md).

use super::common::CodecError;

/// SR04 host→device command opcodes.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum Sr04Cmd {
    GetSamplePeriod = 0x32,
    SetSamplePeriod = 0x33,
    GetEchoDecay = 0x34,
    SetEchoDecay = 0x35,
    MeasureOnce = 0x36,
    StartMeasurementLoop = 0x37,
    StopMeasurementLoop = 0x38,
}

/// SR04 device→host report opcodes.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum Sr04Rpt {
    Data = 0x91,
    SamplePeriod = 0x92,
    EchoDecay = 0x93,
}

/// `echo_time_us` sentinel: no echo received.
pub const ECHO_TIMEOUT: u16 = 0xFFFF;

pub const SAMPLE_PERIOD_DEFAULT_US: u32 = 50_000;
pub const ECHO_DECAY_DEFAULT_US: u16 = 5_000;
pub const ECHO_DECAY_MIN_US: u16 = 4_000;
pub const ECHO_DECAY_MAX_US: u16 = 65_000;

/// A decoded `RPT_DATA` sample.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Sr04Data {
    /// 0x36 single shot (host or SYNC_IN), 0x37 loop sample (ERRATA E3).
    pub source_cmd: u8,
    pub timestamp_us: u64,
    pub echo_time_us: u16,
}

impl Sr04Data {
    pub fn unpack(payload: &[u8]) -> Result<Sr04Data, CodecError> {
        if payload.len() != 11 {
            return Err(CodecError("SR04 data must be 11 bytes"));
        }
        Ok(Sr04Data {
            source_cmd: payload[0],
            timestamp_us: u64::from_le_bytes(payload[1..9].try_into().unwrap()),
            echo_time_us: u16::from_le_bytes(payload[9..11].try_into().unwrap()),
        })
    }

    /// `true` when the sample is the timeout sentinel (no echo).
    pub fn is_timeout(&self) -> bool {
        self.echo_time_us == ECHO_TIMEOUT
    }
}

pub fn pack_sample_period(period_us: u32) -> [u8; 4] {
    period_us.to_le_bytes()
}

pub fn unpack_sample_period(payload: &[u8]) -> Result<u32, CodecError> {
    if payload.len() != 4 {
        return Err(CodecError("sample period must be 4 bytes"));
    }
    Ok(u32::from_le_bytes(payload.try_into().unwrap()))
}

pub fn pack_echo_decay(decay_us: u16) -> [u8; 2] {
    decay_us.to_le_bytes()
}

pub fn unpack_echo_decay(payload: &[u8]) -> Result<u16, CodecError> {
    if payload.len() != 2 {
        return Err(CodecError("echo decay must be 2 bytes"));
    }
    Ok(u16::from_le_bytes(payload.try_into().unwrap()))
}

/// Round-trip echo time → distance in mm; `None` for the timeout sentinel.
///
/// Default speed of sound 343 m/s; with `air_temp_c` uses `c = 331.3 + 0.606·T`.
pub fn distance_mm_from_echo(echo_time_us: u16, air_temp_c: Option<f64>) -> Option<f64> {
    if echo_time_us == ECHO_TIMEOUT {
        return None;
    }
    let c_m_s = match air_temp_c {
        None => 343.0,
        Some(t) => 331.3 + 0.606 * t,
    };
    Some(echo_time_us as f64 * c_m_s / 2000.0)
}
